### DISCUSSION TITLE
Replacing most usage of Ops with Assets - any downside?

### DISCUSSION THREAD

**Discussion by ustulation** [dagster-io/dagster#9687] [Q&A] [answered]
Firstly thanks for `dagster` and thank you for making it open source!

I've been debating with myself if I can just use assets instead of ops + graphs in most scenarios. I'll give an example:

I have some production code that has a group of assets like thus:
```
# prod.py
@dagster.asset
def a0() -> list[str]:
   ...
```
I wanted to use them exactly as they are in many materialisation graphs (with different upstream and downstream deps in each one). One eg. is I just wanted to string a few integration tests to use them in different scenarios. So I came up with this:
```
#prod.py
def asset_factory(**kwargs) -> dagster.AssetDefinition:
    @dagster.asset(**kwargs)
    def a0() -> list[str]:
       ...
    return a0
```
But with that slight indirection of a factory pattern i can make it a part of many distinct graphs:
```
# test_0.py
@dagster.asset
def tear_down_and_setup():
   ...

to_test = prod.asset_factory(non_argument_deps = {tear_down_and_setup.key})

@dagster.asset(ins = { 'coll': dagster.AssetIn(key = to_test.key) })
def tear_down(coll: list[str]):
    ...

assets = dagster.load_assets_from_current_module(group = 'test_0', prefix = 'test_0')
test_0_job = dagster.define_asset_job(name = 'test_0_job', selection = dagster.AssetSelection.assets(*assets))
```
And create `test_1.py` etc just by varying the non-arg-deps for the asset in `prod.py`. All these will show up with their own distinct graphs nicely prefixed and grouped and i can decide to run any job to test the same asset under different setups. I can put all these assets and jobs in a single @dagster.repository called `tests` for example (as they have different groups and prefixes so nothing "collides").

However the asset def. in the tutorials say that assets are supposed to be persistent artifact and all that, which I'm clearly not using them for here - I'm just using them to keep the code uniform (not a mix of ops, graphs and assets)  and to leverage the fact that the assets figure their graphs by themselves. I'm later tearing down the setup so none of these assets are actually persistent beyond the duration of the test. They are only persistent when materialised by the production job (probably in a different dagster repository) with different upstream and downstream deps, not shown here.

**Given all that, my main question is, is there a drawback to using Assets as such (instead of ops + graphs) and if so what are they? Is this an anti-pattern?**

**Comment by sryza** [comment, marked as answer]
Hey @ustulation - I think the asset factory pattern you're describing is entirely reasonable.
